### PR TITLE
Add validation to reject inner columns containing whitespace chars

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

Closes https://github.com/crate/crate/issues/13486

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 1]
```diff
@@ -120,6 +122,8 @@ public DocumentMapper build(MapperService mapperService) {
     private final MetadataFieldMapper[] deleteTombstoneMetadataFieldMappers;
     private final MetadataFieldMapper[] noopTombstoneMetadataFieldMappers;
 
+    private final static Pattern NAME_WITH_WHITESPACE_CHARACTERS = Pattern.compile(".*(?:[\\t\\r\\n]+.*)+");
```
Comment (mkleen) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:125]: I am not a regex-wizard. I wonder if there is a way to do this with a non-capture group ?

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 2]
```diff
@@ -120,6 +122,8 @@ public DocumentMapper build(MapperService mapperService) {
     private final MetadataFieldMapper[] deleteTombstoneMetadataFieldMappers;
     private final MetadataFieldMapper[] noopTombstoneMetadataFieldMappers;
 
+    private static final Pattern NAME_WITH_WHITESPACE_CHARACTERS = Pattern.compile(".*(?:[\\t\\r\\n]+.*)+");
```
Comment (mkleen) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:125]: Is there any way to do this without a none-capturing group ?

File: docs/appendices/release-notes/unreleased.rst [hunk 3]
```diff
@@ -54,6 +54,9 @@ None
 Changes
 =======
 
+- Added validation to reject inner column names containing whitespace characters
```
Comment (matriv) [inline docs/appendices/release-notes/unreleased.rst:57]: Shouldn't this be considered a fix?

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java (hunk 2 above)
Comment (matriv) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:125]: Maybe if you replace all whitespaces with `" "` and check if the strings are the same, but probably isn't going to be faster: `replaceAll("\\p{Zs}+", " ");`

But isn't this problematic if someone uses other weird chars like `\a` or other weird unicode chars?

**Review by matriv — COMMENTED** [review]
Thx! Left a couple of comments.
Also, is it only a problem for object keys (properties)? what about the values?

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java (hunk 2 above)
Comment (mkleen) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:125]: The bell operator `\a` and other weird Unicode chars will not pass the Jackson parser validation. Whitespace characters are valid as JSON keys, but result in weird schemas when used in inner columns.

File: docs/appendices/release-notes/unreleased.rst (hunk 3 above)
Comment (mkleen) [reply docs/appendices/release-notes/unreleased.rst:57]: Right, could also be a fix.

**Comment by mkleen** [comment]
> Thx! Left a couple of comments. Also, is it only a problem for object keys (properties)? what about the values?

Values work because they won't modify the schema.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java (hunk 2 above)
Comment (mfussenegger) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:125]: Could iterate over the characters of the string and check them?

Comment (matriv) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:125]: > Could iterate over the characters of the string and check them?

Could be the faster way, maybe try a jmh benchmark  to check those alternatives.

**Comment by matriv** [comment]
> > Thx! Left a couple of comments. Also, is it only a problem for object keys (properties)? what about the values?
> 
> Values work because they won't modify the schema.

Thx, just double checking!

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java (hunk 2 above)
Comment (mkleen) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:125]: Good idea, i will try that.

Comment (mkleen) [reply server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:125]:
```
Benchmark                                         Mode  Cnt        Score        Error  Units
ColumnNameBenchmark.measure_validate_loop_chars  thrpt   25  7945104.819 ± 518013.577  ops/s
ColumnNameBenchmark.measure_validate_regex       thrpt   25   250635.866 ±   3263.536  ops/s
```
https://gist.github.com/mkleen/bdeff74d3be62890cd5e40a571bb147a
Numbers are clear hear, let's not use a regex ;-)

**Comment by mkleen** [comment]
Ready for another review.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 4]
```diff
@@ -164,6 +175,24 @@ public DocumentMapper(MapperService mapperService, Mapping mapping) {
             .filter(field -> noopTombstoneMetadataFields.contains(field.name())).toArray(MetadataFieldMapper[]::new);
     }
 
+    private static boolean containsInvalidWhitespaceCharacters(String input) {
+        int length = input.length();
+        if (input.length() == 0) {
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:180]: This is not needed, for loop won't execute anyway.

**Review by matriv — APPROVED** [review]
LGTM, Left a comment.

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java [hunk 5]
```diff
@@ -164,6 +175,20 @@ public DocumentMapper(MapperService mapperService, Mapping mapping) {
             .filter(field -> noopTombstoneMetadataFields.contains(field.name())).toArray(MetadataFieldMapper[]::new);
     }
 
+    private static boolean containsInvalidWhitespaceCharacters(String input) {
+        for (int i = 0; i < input.length(); i++) {
+            char currentChar = input.charAt(i);
+            switch (currentChar) {
```
Comment (matriv) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentMapper.java:181]: minor, but you don't need to assign it to a var.

**Pull request merged**